sn_cli/src/subcommands/folders.rs

```diff
@@ -0,0 +1,315 @@
+
+use super::files::{download_file, upload_files, ChunkManager, UploadedFile, UPLOADED_FILES};
+
+use sn_client::{
+    Client, FilesApi, FolderEntry, FoldersApi, WalletClient, BATCH_SIZE, MAX_UPLOAD_RETRIES,
+};
+
+use sn_protocol::storage::{Chunk, ChunkAddress, RegisterAddress};
+use sn_transfers::HotWallet;
+
+use clap::Parser;
+use color_eyre::{
+    eyre::{bail, eyre},
+    Result,
+};
+use std::{
+    collections::BTreeMap,
+    fs::create_dir_all,
+    path::{Path, PathBuf},
+};
+use tokio::task::JoinSet;
+use walkdir::WalkDir;
+
+#[derive(Parser, Debug)]
+pub enum FoldersCmds {
+    Upload {
+        /// The location of the file(s) to upload for creating the folder on the network.
+        ///
+        /// Can be a file or a directory.
+        #[clap(name = "path", value_name = "PATH")]
+        path: PathBuf,
+        /// The batch_size to split chunks into parallel handling batches
+        /// during payment and upload processing.
+        #[clap(long, default_value_t = BATCH_SIZE, short='b')]
+        batch_size: usize,
+        /// Should the file be made accessible to all. (This is irreversible)
+        #[clap(long, name = "make_public", default_value = "false", short = 'p')]
+        make_public: bool,
+        /// The retry_count for retrying failed chunks
+        /// during payment and upload processing.
+        #[clap(long, default_value_t = MAX_UPLOAD_RETRIES, short = 'r')]
+        max_retries: usize,
+    },
+    Download {
+        /// The hex address of a folder.
+        #[clap(name = "address")]
+        folder_addr: String,
+        /// The name to apply to the downloaded folder.
+        #[clap(name = "target folder name")]
+        folder_name: String,
+        /// The batch_size for parallel downloading
+        #[clap(long, default_value_t = BATCH_SIZE , short='b')]
+        batch_size: usize,
+    },
+}
+
+pub(crate) async fn folders_cmds(
+    cmds: FoldersCmds,
+    client: &Client,
+    root_dir: &Path,
+    verify_store: bool,
+) -> Result<()> {
+    match cmds {
+        FoldersCmds::Upload {
+            path,
+            batch_size,
+            max_retries,
+            make_public,
+        } => {
+            upload_files(
+                path.clone(),
+                make_public,
+                client,
+                root_dir.to_path_buf(),
+                verify_store,
+                batch_size,
+                max_retries,
+            )
+            .await?;
+
+            let mut chunk_manager = ChunkManager::new(root_dir);
+            chunk_manager.chunk_path(&path, true, make_public)?;
+
+            let mut folders = build_folders_hierarchy(&path, client, root_dir)?;
+
+            // add chunked files to the corresponding Folders
+            for chunked_file in chunk_manager.iter_chunked_files() {
+                if let (Some(file_name), Some(parent)) = (
+                    chunked_file.file_name.to_str(),
```

Comment:
Can we store the `OsString` directly into the registers? Why do we convert it into str here? Can this lead to OS related anomalies?

Reply:
I'll give it a try

Reply:
there is a problem with serialisation for wasm, I'll keep it for now, it'll go away once we move metadata out from Register anyways.